botocore/configprovider.py

```diff
@@ -836,3 +858,159 @@ def provide(self):
+                )
+                return endpoint_url
+
+        logger.debug('No configured endpoint found.')
+        return None
+
+    def _get_snake_case_service_id(self, client_name):
+        # Use lookups to get the serice ID without loading the service data
+        # file. `client_name`` refers to the name used to instantiate a client
+        # through botocore.session.Session.create_client (parameter name there
+        # is `service_name`). This client name is generally the hyphenized
+        # service ID.  We need to look up services that have been renamed
+        # (SERVICE_NAME_ALIASES) as well as services that do not use
+        # the service ID as their data directory name
+        # (CLIENT_NAME_TO_HYPHENIZED_SERVICE_ID_OVERRIDES).
+        client_name = utils.SERVICE_NAME_ALIASES.get(client_name, client_name)
+        hyphenized_service_id = (
+            utils.CLIENT_NAME_TO_HYPHENIZED_SERVICE_ID_OVERRIDES.get(
+                client_name, client_name
+            )
+        )
+        return hyphenized_service_id.replace('-', '_')
+
+    def _get_service_env_var_name(self):
+        transformed_service_id_env = self._get_snake_case_service_id(
+            self._client_name
+        ).upper()
+        return f'AWS_ENDPOINT_URL_{transformed_service_id_env}'
+
+    def _get_services_config(self):
+        if 'services' not in self._scoped_config:
+            return {}
+
+        section_name = self._scoped_config['services']
+        services_section = self._full_config.get('services', {}).get(
+            section_name
+        )
+
+        if not services_section:
+            error_msg = (
+                f'The profile is configured to use the services '
+                f'section but the "{section_name}" services '
+                f'configuration does not exist.'
+            )
+            raise InvalidConfigError(error_msg=error_msg)
+
+        return services_section
+
+    def _get_endpoint_url_config_service(self):
+        snakecase_service_id = self._get_snake_case_service_id(
+            self._client_name
+        ).lower()
+        return (
+            self._get_services_config()
+            .get(snakecase_service_id, {})
+            .get('endpoint_url')
+        )
+
+    def _get_endpoint_url_config_global(self):
+        return self._scoped_config.get('endpoint_url')
+
+    def _get_endpoint_url_environment_service(self):
+        return EnvironmentProvider(
+            name=self._get_service_env_var_name(), env=self._environ
+        ).provide()
+
+    def _get_endpoint_url_environment_global(self):
+        return EnvironmentProvider(
+            name='AWS_ENDPOINT_URL', env=self._environ
+        ).provide()
+
+    def __repr__(self):
+        return (
+            f'ConfiguredEndpointProvider('
+            f'full_config={self._full_config}, '
+            f'scoped_config={self._scoped_config}, '
+            f'client_name={self._client_name}, '
+            f'environ={self._environ})'
```

Comment:
This seems like it could be _really_ big. Do we want need all config and environ variables in the repr?

Reply:
I think was done to match parity with the other config providers (e.g. it is done here for the `EnvironmentProvider`: https://github.com/boto/botocore/blob/f2d4384ff9213d732bcb88dc4c8f3bf292e7e76f/botocore/configprovider.py#L781-L782). Otherwise, I don't think there was any other reason.

Thinking about it more @kdaily do you know if the repr gets logged anywhere in the debug logs? It could be problematic if say the env var or the config had credentials in them as we would not want those logged.

Reply:
The only debug logs added for this feature are to output one of the [names of the providers](https://github.com/boto/botocore/pull/2973/files#diff-1818b8b628e5570486a0305e24ca6036a980fae9b2b73125b3298c7d02163313R871-R876) and the client name.

- looking for an endpoint from a provider [here](https://github.com/boto/botocore/pull/2973/files#diff-1818b8b628e5570486a0305e24ca6036a980fae9b2b73125b3298c7d02163313R925)
- found an endpoint from a provider [here](https://github.com/boto/botocore/pull/2973/files#diff-1818b8b628e5570486a0305e24ca6036a980fae9b2b73125b3298c7d02163313R934)

Reply:
It does feel that it would be risky to print the environment or config representation by default for that reason.

Reply:
Removed the repr.